### PR TITLE
Fix bug in ignore include paths for sym links

### PR DESCRIPTION
**Pull request by jjwilke**

### CODE REVIEW

File: bin/clang/astVisitor.cc [hunk 1]
```diff
@@ -212,12 +212,23 @@ CompilerGlobals::setup(clang::CompilerInstance* CI)
...
+      //make sure the full paths (following symlinks)
+      //are included for checking
+      const char* fullpath_cstr = realpath(path.c_str(), fullpathBuffer);
+      if (fullpath_cstr){
+        std::string fullpath = fullpath_cstr;
+        //if we follow symlinks or remove rel path and is different
+        if (fullpath != path){
+          includePaths.push_back(fullpath);
+        }
+      }
```
Comment (calewis) [inline bin/clang/astVisitor.cc:231]: The way that I read realpath as working, means that we should just drop line 221 and just always push_back fullpath?? And error if fullpath_cstr is null? 

Even better we can emplace_back fullpath_cstr and not require the explicit conversion. Hopefully I didn't check this.

Comment (jjwilke) [reply bin/clang/astVisitor.cc:231]: Just for safety's sake, if there is any checking of paths done. I want to make sure we catch either the symlink path or the fundamental path.

File: bin/clang/astVisitor.cc [hunk 2]
```diff
@@ -212,12 +212,22 @@ CompilerGlobals::setup(clang::CompilerInstance* CI)
...
     std::string path;
     while (std::getline(sstr, path, ':'))
     {
-      includePaths.push_back(path);
+      //make sure the full paths (following symlinks)
+      //are included for checking
+      const char* fullpath_cstr = realpath(path.c_str(), fullpathBuffer);
+      if (fullpath_cstr){
+        realSystemIncludePaths.push_back(fullpath_cstr);
+      } else {
```
Comment (calewis) [inline bin/clang/astVisitor.cc:226]: Can `realpath` fail while the path is valid?  Part of my original concern I wanted addressed was the question that if `realpath` fails do we just have an error?

File: bin/clang/astVisitor.cc [hunk 3]
```diff
@@ -212,12 +212,22 @@ CompilerGlobals::setup(clang::CompilerInstance* CI)
...
     std::istringstream sstr(sref.str());
     std::string path;
     while (std::getline(sstr, path, ':'))
     {
-      includePaths.push_back(path);
+      //make sure the full paths (following symlinks)
+      //are included for checking
+      const char* fullpath_cstr = realpath(path.c_str(), fullpathBuffer);
+      if (fullpath_cstr){
+        realSystemIncludePaths.push_back(fullpath_cstr);
```
Comment (calewis) [inline bin/clang/astVisitor.cc:225]: `emplace_back`?

File: bin/clang/astVisitor.cc [hunk 4]
```diff
@@ -212,12 +212,22 @@ CompilerGlobals::setup(clang::CompilerInstance* CI)
 
 
   if (includeListOpt.getNumOccurrences()){
+    char fullpathBuffer[1024];
     StringRef sref(includeListOpt);
     std::istringstream sstr(sref.str());
```
Comment (calewis) [inline bin/clang/astVisitor.cc:217]: Don't need a StringRef here.  Just use includeListOpt.getValue().  Avoids a copy.

File: bin/clang/astVisitor.cc (hunk 3 above)
Comment (jjwilke) [reply bin/clang/astVisitor.cc:225]: Changed

File: bin/clang/astVisitor.cc (hunk 4 above)
Comment (jjwilke) [reply bin/clang/astVisitor.cc:217]: Changed

File: bin/clang/astVisitor.cc (hunk 2 above)
Comment (jjwilke) [reply bin/clang/astVisitor.cc:226]: Yeah, I guess it's probably safer to error out. I haven't encountered a system so far that realpath fails on.

File: bin/clang/astVisitor.cc [hunk 5]
```diff
@@ -212,12 +212,21 @@ CompilerGlobals::setup(clang::CompilerInstance* CI)
 
 
   if (includeListOpt.getNumOccurrences()){
-    StringRef sref(includeListOpt);
-    std::istringstream sstr(sref.str());
+    char fullpathBuffer[1024];
+    //StringRef sref(includeListOpt);
```
Comment (calewis) [inline bin/clang/astVisitor.cc:216]: Remove comment before merge.

File: bin/clang/astVisitor.cc [hunk 6]
```diff
@@ -414,47 +423,51 @@ SkeletonASTVisitor::shouldVisitDecl(VarDecl* D)
...
+      return false;
+    }
+  }
+
+  //not an ignored header - good to go
+  return true;
+}
+
+bool
+SkeletonASTVisitor::isSystemHeader(SourceLocation sourceLoc, const std::string& header_path)
```
Comment (calewis) [inline bin/clang/astVisitor.cc:436]: Just take a SourceLocation and get the header from that.

File: bin/clang/astVisitor.cc [hunk 7]
```diff
@@ -414,47 +423,51 @@ SkeletonASTVisitor::shouldVisitDecl(VarDecl* D)
...
+  const char* fullpath_cstr = realpath(header_path.c_str(), fullpathBuffer);
+  if (fullpath_cstr){
+    std::string fullpath(fullpath_cstr);
+    if (validHeaders_.empty()){
+      //we have not been explicitly given a list of valid headers
+      //just ignore all headers in default system paths
+      //we are inside a header
+      //this is a system header - skip it
+      for (auto&& path : CompilerGlobals::realSystemIncludePaths){
+        if (path.size() < fullpath.size()){
```
Comment (calewis) [inline bin/clang/astVisitor.cc:448]: Use the methods in StringRef to clean this code up.

File: bin/clang/astVisitor.h [hunk 8]
```diff
@@ -933,6 +933,13 @@ class SkeletonASTVisitor : public clang::RecursiveASTVisitor<SkeletonASTVisitor>
    */
   const clang::Decl* getOriginalDeclaration(clang::VarDecl* vd);
 
+  /**
+    @param path The path to a header to check if included in a systme folder
+    @param loc  The source location the include came from
+    @return Whether this is a "system" header that should be ignored in source-to-source
+  */
+  bool isSystemHeader(clang::SourceLocation loc, const std::string& path);
```
Comment (calewis) [inline bin/clang/astVisitor.h:941]: Change name to `isInSystemHeader` mirrors https://clang.llvm.org/doxygen/classclang_1_1FullSourceLoc.html#a98ffaccf4e6822a7ffced0d7f6b4aee1

File: bin/clang/clangGlobals.h [hunk 9]
```diff
@@ -240,7 +240,11 @@ struct CompilerGlobals {
     return ci->getSema();
   }
 
-  static std::list<std::string> includePaths;
+  /**
+   This contains the list of systme include root paths.
```
Comment (calewis) [inline bin/clang/clangGlobals.h:244]: Typo

File: bin/clang/astVisitor.cc [hunk 10]
```diff
@@ -414,47 +423,51 @@ SkeletonASTVisitor::shouldVisitDecl(VarDecl* D)
 
   bool useAllHeaders = false;
```
Comment (calewis) [inline bin/clang/astVisitor.cc:424]: Unless I am missing something useAllHeaders can go away.

File: bin/clang/astVisitor.cc [hunk 11]
```diff
@@ -414,47 +423,51 @@ SkeletonASTVisitor::shouldVisitDecl(VarDecl* D)
...
-            }
-            if (substr_match){
-              match = true;
+    if (isSystemHeader(startLoc, ploc.getFilename())){
+      return false;
+    }
+  }
+
+  //not an ignored header - good to go
+  return true;
```
Comment (calewis) [inline bin/clang/astVisitor.cc:432]: I think last chunk of this function could be rewritten as: 
```cpp
return headerLoc.isValid() ? !isSystemHeader(/* */) : true;
```
or
```cpp
if(headerLoc.isValid() && isSystemHeader(startLoc, ploc.getFilename())){
   return false;
}

return true;
```

**Pull request merged**